### PR TITLE
[SPARK-42611][SQL] Insert char/varchar length checks for inner fields during resolution

### PR DESCRIPTION
**Pull request by aokolnychyi**
### What changes were proposed in this pull request?

This PR adds  char/varchar length checks for inner fields during resolution when struct fields are reordered.

### Why are the changes needed?

These checks are needed to handle nested char/varchar columns correctly. Prior to this change, we would lose the raw type information when constructing nested attributes. As a result, we will not insert proper char/varchar length checks.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This PR comes with tests that would previously fail.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 1]
```diff
@@ -926,4 +926,56 @@ class DSV2CharVarcharTestSuite extends CharVarcharTestSuite
       }
     }
   }
+
+  test("SPARK-42611: check char/varchar length in reordered nested structs") {
```
Comment (aokolnychyi) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:930]: We would skip char/varchar length checks in all these scenarios prior to this change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 2]
```diff
@@ -139,33 +139,46 @@ object TableOutputResolver {
...
+      expectedCol: Attribute,
       expectedType: StructType,
-      expectedName: String,
       conf: SQLConf,
       addError: String => Unit,
       colPath: Seq[String]): Option[NamedExpression] = {
     val fields = inputType.zipWithIndex.map { case (f, i) =>
       Alias(GetStructField(input, i, Some(f.name)), f.name)()
     }
-    val reordered = reorderColumnsByName(fields, expectedType.toAttributes, conf, addError, colPath)
```
Comment (aokolnychyi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:150]: Whenever we call `expectedType.toAttributes`, we ignore the raw type information associated with the attribute. As a result, we never added length checks when nested fields were resolved by name.

**Comment by aokolnychyi** [comment]
cc @huaxingao @cloud-fan @dongjoon-hyun @sunchao @viirya

**Review by dongjoon-hyun — COMMENTED** [review]
cc @gengliangwang , too

**Review by dongjoon-hyun — COMMENTED** [review]
According to the JIRA Affected Version, is this PR only applicable for Apache Spark 3.4?
Actually, [SPARK-36498](https://issues.apache.org/jira/browse/SPARK-36498) is added at Spark 3.3.0.

**Comment by aokolnychyi** [comment]
@dongjoon-hyun, I overlooked. I corrected the JIRA description and added 3.3.3 as well but feel free to adjust the list of versions as needed.

**Comment by dongjoon-hyun** [comment]
Thank you!

**Comment by dongjoon-hyun** [comment]
Could you rebase to the `master` branch?

**Comment by cloud-fan** [comment]
This is a good catch! I think the fix can be simpler: we encode the raw type information in the root `StructField`, so we don't need to decode it recursively. We can restore the raw type at the very beginning:
```
def resolveOutputColumns(
    ...,
    expected: Seq[Attribute],
    ...) {
    val actualExpectedCols = expected.map { attr =>
        attr.withDataType(restore(attr.metadata))
    }
}
```

**Comment by aokolnychyi** [comment]
@cloud-fan, are thinking of passing raw types all the way? I actually considered that but my worry was we will have to modify `checkField` casting logic as it seems we can't use the raw type there.

```
  private def checkField(
      tableAttr: Attribute,
      queryExpr: NamedExpression,
      byName: Boolean,
      conf: SQLConf,
      addError: String => Unit,
      colPath: Seq[String]): Option[NamedExpression] = {

    val storeAssignmentPolicy = conf.storeAssignmentPolicy
    lazy val outputField = if (tableAttr.dataType.sameType(queryExpr.dataType) &&
      tableAttr.name == queryExpr.name &&
      tableAttr.metadata == queryExpr.metadata) {
      Some(queryExpr)
    } else {
      val casted = storeAssignmentPolicy match {
        case StoreAssignmentPolicy.ANSI =>
          val cast = Cast(queryExpr, tableAttr.dataType, Option(conf.sessionLocalTimeZone),
            ansiEnabled = true)
          cast.setTagValue(Cast.BY_TABLE_INSERTION, ())
          checkCastOverflowInTableInsert(cast, colPath.quoted)
        case StoreAssignmentPolicy.LEGACY =>
          Cast(queryExpr, tableAttr.dataType, Option(conf.sessionLocalTimeZone),
            ansiEnabled = false)
        case _ =>
          Cast(queryExpr, tableAttr.dataType, Option(conf.sessionLocalTimeZone))
      }
      val exprWithStrLenCheck = if (conf.charVarcharAsString) {
        casted
      } else {
        CharVarcharUtils.stringLengthCheck(casted, tableAttr)
      }
      // Renaming is needed for handling the following cases like
      // 1) Column names/types do not match, e.g., INSERT INTO TABLE tab1 SELECT 1, 2
      // 2) Target tables have column metadata
      Some(Alias(exprWithStrLenCheck, tableAttr.name)(explicitMetadata = Some(tableAttr.metadata)))
    }

    storeAssignmentPolicy match {
      case StoreAssignmentPolicy.LEGACY =>
        outputField

      case StoreAssignmentPolicy.STRICT | StoreAssignmentPolicy.ANSI =>
        // run the type check first to ensure type errors are present
        val canWrite = DataType.canWrite(
          queryExpr.dataType, tableAttr.dataType, byName, conf.resolver, colPath.quoted,
          storeAssignmentPolicy, addError)
        if (queryExpr.nullable && !tableAttr.nullable) {
          addError(s"Cannot write nullable values to non-null column '${colPath.quoted}'")
          None

        } else if (!canWrite) {
          None

        } else {
          outputField
        }
    }
  }
```

**Comment by aokolnychyi** [comment]
@cloud-fan, I think passing raw types from the start would be simper and more efficient but we will have to modify `checkField` to work with raw types. Let me know if I got you correctly.

**Comment by cloud-fan** [comment]
Yea, we will have to modify `checkField`. We can replace char/varchar type with string type in `checkField` so that most code can be unchanged, and only use the original type for length check related logic.

**Comment by aokolnychyi** [comment]
Sounds good, I'll make the change today.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side. Thank you for the update.
I believe we need @cloud-fan 's final sign-off with the last commit.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 3]
```diff
@@ -926,4 +926,56 @@ class DSV2CharVarcharTestSuite extends CharVarcharTestSuite
...
+    Seq("CHAR(5)", "VARCHAR(5)").foreach { typ =>
+      withTable("t") {
+        sql(s"CREATE TABLE t(s STRUCT<n_c:$typ,n_i:INT>) USING $format")
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:933]: extremely nit: shall we have an empty space in between of `n_c:$typ,` and `n_i:INT` in this line and also in the `CREATE TABLE` statements below?

**Review by huaxingao — APPROVED** [review]

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 4]
```diff
@@ -40,16 +40,20 @@ object TableOutputResolver {
...
+
     val errors = new mutable.ArrayBuffer[String]()
     val resolved: Seq[NamedExpression] = if (byName) {
-      reorderColumnsByName(query.output, expected, conf, errors += _)
+      reorderColumnsByName(query.output, actualExpectedCols, conf, errors += _)
     } else {
       if (expected.size > query.output.size) {
         throw QueryCompilationErrors.cannotWriteNotEnoughColumnsToTableError(
           tableName, expected, query)
       }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:54]: Should we put `actualExpectedCols` in error?
```suggestion
      if (expected.size > query.output.size) {
        throw QueryCompilationErrors.cannotWriteNotEnoughColumnsToTableError(
          tableName, actualExpectedCols, query)
      }
```

Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:54]: @viirya, there are a few places like this. I will update all of them even though we just use names in errors.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala (hunk 3 above)
Comment (aokolnychyi) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:933]: Will do, I overlooked the formatting in this class.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala (hunk 4 above)
Comment (aokolnychyi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala (hunk 3 above)
Comment (aokolnychyi) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala]: Done.

**Review by cloud-fan — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @aokolnychyi , @cloud-fan , @huaxingao , @viirya !

**Pull request closed without merging**